external/llvm-project/mlir/lib/Dialect/Math/Transforms/ExtendToLLVMTypes.cpp

```diff
@@ -0,0 +1,175 @@
+//===- LegalizeToF32.cpp - Legalize functions on small floats ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements legalizing math operations on small floating-point
+// types through arith.extf and arith.truncf.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Math/Transforms/Passes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SetVector.h"
+
+namespace mlir::math {
+#define GEN_PASS_DEF_MATHEXTENDTOLLVMTYPES
+#include "mlir/Dialect/Math/Transforms/Passes.h.inc"
+} // namespace mlir::math
+
+using namespace mlir;
+
+namespace {
+struct ExtendToLLVMTypesRewritePattern final : ConversionPattern {
+  ExtendToLLVMTypesRewritePattern(TypeConverter &converter,
+                                  MLIRContext *context)
+      : ConversionPattern(converter, MatchAnyOpTypeTag{}, 1, context) {}
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override;
+};
+
+struct ExtendToLLVMTypesPass
+    : mlir::math::impl::MathExtendToLLVMTypesBase<ExtendToLLVMTypesPass> {
+  using math::impl::MathExtendToLLVMTypesBase<
+      ExtendToLLVMTypesPass>::MathExtendToLLVMTypesBase;
+
+  void runOnOperation() override;
+};
+} // namespace
+
+/// Map strings to float types. This function is here because no one else needs
+/// it yet, feel free to abstract it out.
+static std::optional<FloatType> parseFloatType(MLIRContext *ctx,
```

Comment:
I'd go looking for some sort of utilities library or somewhere under `IR/`.

Or, really, `Arith/Utils` might work, since `math` depends on `arith` last I checked.